internals/search_queries.py

```diff
@@ -0,0 +1,198 @@
+
+import logging
+
+from framework import utils
+from internals import models
+
+
+def single_field_query_async(
+    field_name, operator, val, limit=None):
+  """Create a query for one Feature field and run it, returning a promise."""
+  field = QUERIABLE_FIELDS.get(field_name.lower())
+  if field is None:
+    logging.info('Ignoring field name %r', field_name)
+    return []
+  # TODO(jrobbins): support sorting by any fields of other model classes.
+  query = models.Feature.query()
+  # Note: We don't exclude deleted features, that's done by process_query.
+
+  # TODO(jrobbins): Handle ":" operator as substrings for text fields.
+  if (operator == '='):
+    query = query.filter(field == val)
+  elif (operator == '<='):
+    query = query.filter(field <= val)
+  elif (operator == '<'):
+    query = query.filter(field < val)
+  elif (operator == '>='):
+    query = query.filter(field >= val)
+  elif (operator == '>'):
+    query = query.filter(field > val)
+  elif (operator == '!='):
+    query = query.filter(field != val)
+  else:
+    raise ValueError('Unexpected query operator: %r' % operator)
+
+  keys_promise = query.fetch_async(keys_only=True, limit=limit)
+  return keys_promise
+
+
+def total_order_query_async(sort_spec):
+  """Create a query promise for all Feature IDs sorted by sort_spec."""
+  # TODO(jrobbins): Support multi-column sort.
+  descending = False
+  if sort_spec.startswith('-'):
+    descending = True
+    sort_spec = sort_spec[1:]
+  field = SORTABLE_FIELDS.get(sort_spec.lower())
+  if field is None:
+    logging.info('Ignoring sort field name %r', sort_spec)
+    return []
+
+  # Some special cases are implemented as python functions.
+  if callable(field):
+    return field(descending)
+
+  if descending:
+    field = -field
+  # TODO(jrobbins): support sorting by any fields of other model classes.
+  query = models.Feature.query().order(field)
+
+  keys_promise = query.fetch_async(keys_only=True)
+  return keys_promise
+
+
+def sorted_by_pending_request_date(descending):
+  """Return feature_ids of pending approvals sorted by request date."""
+  query = models.Approval.query(
+      models.Approval.state == models.Approval.REVIEW_REQUESTED)
+  if descending:
+    query = query.order(-models.Approval.set_on)
+  else:
+    query = query.order(models.Approval.set_on)
+
+  pending_approvals = query.fetch(projection=['feature_id'])
+  feature_ids = utils.dedupe(pa.feature_id for pa in pending_approvals)
+  return feature_ids
+
+
+def sorted_by_review_date(descending):
+  """Return feature_ids of reviewed approvals sorted by last review."""
+  query = models.Approval.query(
+      models.Approval.state.IN(models.Approval.FINAL_STATES))
+  if descending:
+    query = query.order(-models.Approval.set_on)
+  else:
+    query = query.order(models.Approval.set_on)
+  recent_approvals = query.fetch(projection=['feature_id'])
+
+  feature_ids = utils.dedupe(ra.feature_id for ra in recent_approvals)
+  return feature_ids
```

Comment:
These functions seem very similar. Would there be utility in merging them and making them reusable?

Reply:
Good point.  I made a helper function to avoid repeating similar code.